hal/src/rtl872x/interrupts_hal.cpp

```diff
@@ -257,24 +258,39 @@ int hal_interrupt_set_direct_handler(IRQn_Type irqn, hal_interrupt_direct_handle
         return 1;
     }
 
+    int error = 0;
     int32_t state = HAL_disable_irq();
-    volatile uint32_t* isrs = (volatile uint32_t*)&link_ram_interrupt_vectors_location;
 
     if (handler == nullptr && (flags & HAL_INTERRUPT_DIRECT_FLAG_RESTORE)) {
-        // Restore
-        // HAL_Core_Restore_Interrupt(irqn);
+        // Restore old handler only if one was backed up
+        uint32_t old_handler = hal_interrupts_handler_backup[IRQN_TO_IDX(irqn)];
+        if (old_handler) {
+            __NVIC_SetVector(irqn, (uint32_t)old_handler);
+            hal_interrupts_handler_backup[IRQN_TO_IDX(irqn)] = 0;
+        }
     } else {
-        isrs[IRQN_TO_IDX(irqn)] = (uint32_t)handler;
+        // If there is currently a handler backup: Return error
+        if (hal_interrupts_handler_backup[IRQN_TO_IDX(irqn)]) {
+            error = SYSTEM_ERROR_NOT_SUPPORTED;
```

Comment:
It might actually simplify the logic if I use `goto` to bail out rather than have to check for error conditions? I cant `return` here because I need to re-enable interrupts.

Reply:
Use `CHECK_TRUE`/`CHECK_FALSE` then, much better than `goto`, so:

```c++
CHECK_FALSE(hal_interrupts_handler_backup[IRQN_TO_IDX(irqn)], SYSTEM_ERROR_ALREADY_EXISTS);
```

Reply:
Ah yes, those macros are handy. Unfortunately I need to re-enable interrupts before exiting. ill just do that and return `SYSTEM_ERROR_ALREADY_EXISTS`. I think thats cleaner than adding a return variable and checking it, etc.